### PR TITLE
xfce4-14: 4.14.pre2 -> 4.14

### PR DESCRIPTION
**Pull request by volth**
XFCE 4.14 released

### CODE REVIEW

**Comment by jonringer** [comment]
looked into the xfburn failure, it's unable to get libexo-1.so from the exo package. According to [the exo homepage](http://www.linuxfromscratch.org/blfs/view/cvs/xfce/exo.html) they export libexo-1.so, unless it was recently retracted, it should be there. However, I was unsucessful to toggle it on or see an option for it with the ./configure step.

**Comment by jonringer** [comment]
Looked into it more, libexo-1.so, is exported by xfce4-12.exo, and according to the homepage it still requires 4-12, I don't think xfburn should be moved under 4-14 until has support for the new upstream packages.

**Comment by volth** [comment]
> I don't think xfburn should be moved under 4-14

Let's make it in the next PR.
There is `xfburn` to delete and [some packages](https://github.com/NixOS/nixpkgs/compare/af34b573a3f7efa986f53cd223b5201fd5ee680a..49526093f9b205220322079a9a654008bd01cab2) to add.
Then make 4.14 default instead of 4.12.
If we will do everything in a single PR it will get old and bloated and no one would dare to merge it.

**Comment by jonringer** [comment]
sounds reasonable. smaller changes are easier to manage.

**Comment by jtojnar** [comment]
Almost every GTK 3 program should have `wrapGAppsHook` in its `nativeBuildInputs`. Especially when they use `GtkFileChooserDialog`, since the app will crash when GSettings schemas are missing.

Also, with exo, you should be able  to move `docbook_xsl` to `nativeBuildInputs`, making that part of `postPatch` section redundant. I opened an [issue](https://bugzilla.xfce.org/show_bug.cgi?id=15825) about the other part in line with https://github.com/NixOS/nixpkgs/issues/36468.

**Comment by jonringer** [comment]
erroneously mentioned this PR, i meant 66559, not 66559.

**Comment by volth** [comment]
related: https://github.com/NixOS/nixpkgs/pull/66568 "xfce4-14.xfdashboard: init at 0.7.5"

**Comment by worldofpeace** [comment]
I'm thinking the best way to go with this is to 
* get the packaging here pretty good first
* finish the module https://github.com/NixOS/nixpkgs/pull/37594
* deprecate and remove the old package set (https://github.com/NixOS/nixpkgs/pull/37594#issuecomment-511116677 https://github.com/NixOS/nixpkgs/pull/37594#issuecomment-511126923)

**Comment by worldofpeace** [comment]
I gave this a rough start based off your feedback @jtojnar.

XFCE 14 changelogs are here: https://xfce.org/download/changelogs/4.14

We should probably consolidate all optional gtk2 deps
> Exo has an optional dependency on GTK2.
Garcon has an optional dependency on GTK2.
Libxfce4ui has an optional dependency on GTK2.
Xfce4-panel has an optional dependency on GTK2.

And I think some things have optional `gobject-introspection` and or `vala` support.

Lastly `xfburn` fails to build.

**Comment by volth** [comment]
> We should probably consolidate all optional gtk2 deps
>> `Xfce4-panel` has an optional dependency on GTK2.

There are plenty of panel plugins (both official from xfce.org and 3rd party's like `dockbarx`) inherited from 4.12 which still need GTK2.
They work well in 4.14.

**Comment by jtojnar** [comment]
> There are plenty of panel plugins (both official from xfce.org and 3rd party's like `dockbarx`) inherited from 4.12 which still need GTK2.

Weird that there are no symbol clashes. Do we have any of them in nixpkgs?

> I gave this a rough start based off your feedback @jtojnar.
> (pushing to the branch should be fine now)

Looks good so far. 

> And I think some things have optional `gobject-introspection` and or `vala` support.

If you feel like adding it… But I would not make it a priority. Nothing probably uses it at the moment.

-----

LOL, they have an app called gigolo.

**Comment by volth** [comment]
> > There are plenty of panel plugins (both official from xfce.org and 3rd party's like `dockbarx`) inherited from 4.12 which still need GTK2.
> 
> Weird that there are no symbol clashes. Do we have any of them in nixpkgs?

https://github.com/nixos/nixpkgs/tree/master/pkgs/desktops/xfce/panel-plugins/

**Comment by worldofpeace** [comment]
> @jtojnar 
> And I think some things have optional gobject-introspection and or vala support.
> If you feel like adding it… But I would not make it a priority. Nothing probably uses it at the moment.

Yeah, I don't think there's an xfce panel plugins that use it.

> @jtojnar 
> LOL, they have an app called gigolo.

Yeah :rofl: it seems they like their French words. Even if historically French.

> @volth 
> There are plenty of panel plugins (both official from xfce.org and 3rd party's like dockbarx) inherited from 4.12 which still need GTK2.

No intention to remove anything that needs the optionally supported GTK2 as we have things that need it.

**Comment by worldofpeace** [comment]
`xfburn` fails like:

```
checking pkg-config is at least version 0.9.0... yes
checking for exo-0.3 >= 0.3.4... not found
checking for pkg-config... (cached) /nix/store/zqafyj7by0vpy76ilj4vbfancbcbq22j-pkg-config-0.29.2/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for exo-1 >= 0.3.100... not found
*** The required package exo-1 was not found on your system.
*** Please install exo-1 (atleast version 0.3.100) or adjust
*** the PKG_CONFIG_PATH environment variable if you
*** installed the package in a nonstandard prefix so that
*** pkg-config is able to find it.
```

around https://github.com/xfce-mirror/xfburn/blob/xfburn-0.5.5/configure.ac.in#L40

**Comment by worldofpeace** [comment]
Ah missed @jonringer comment https://github.com/NixOS/nixpkgs/pull/66519#issuecomment-520572072

> Looked into it more, libexo-1.so, is exported by xfce4-12.exo, and according to the homepage it still requires 4-12, I don't think xfburn should be moved under 4-14 until has support for the new upstream packages.

I think we'll handle that in the 
> deprecate and remove the old package set (#37594 (comment) #37594 (comment))

step.

File: pkgs/desktops/xfce4-14/thunar/default.nix [hunk 1]
```diff
@@ -1,5 +1,5 @@
 { mkXfceDerivation, docbook_xsl, exo, gdk-pixbuf, gtk3, libgudev ? null
-, libnotify ? null, libX11, libxfce4ui, libxfce4util, libxslt, xfconf, gobject-introspection }:
+, libnotify ? null, libX11, libxfce4ui, libxfce4util, libxslt, xfconf, gobject-introspection, gvfs }:
```
Comment (volth) [inline pkgs/desktops/xfce4-14/thunar/default.nix]: those `? null` across `xfce4-14` could be cleaned up, I see no purpose to have fallback value for params which are always filled by `callPackage`

Comment (worldofpeace) [reply pkgs/desktops/xfce4-14/thunar/default.nix]: Agreed, done in https://github.com/NixOS/nixpkgs/pull/66519/commits/019d33d5f02d560fd19acd8bda9ee6a954356083

**Comment by worldofpeace** [comment]
I'm not sure there's much more for me to complete here.

If you can @volth it should be ready for testing.

**Comment by volth** [comment]
I'd say it is slower (compared to the system with only first my commit).
Opening setting dialogs takes visible time, changing themes makes cpu heat (fun is noisy).

But I did not do the deep comparison

Do the wrapper load extra libraries or run daemons?

**Comment by worldofpeace** [comment]
> I'd say it is slower (compared to the system with only first my commit).
> Opening setting dialogs takes visible time, changing themes makes cpu heat (fun is noisy).

Most applications are now started with a wrapper, but the footprint on startup time has to be almost nothing.


> But I did not do the deep comparison
> Do the wrappers load extra libraries or run daemons?

These commits enabled features that we could enable easily because they should be in the desktop clousure

* https://github.com/NixOS/nixpkgs/pull/66519/commits/92841bea27d2fd8fb1acd33c0b67bd94fc340e1a
* https://github.com/NixOS/nixpkgs/pull/66519/commits/822325ed664735a5313960eff74547b4b2879452
* https://github.com/NixOS/nixpkgs/pull/66519/commits/3b88e5424e2da91a505a5479eb05add122df8bc9

Other than that, some apps are actually properly wrapped with `GIO_EXTRA_MODULES` for `gvfs` and `glib-networking`. So the module would need `services.gnome3.glib-networking` and `services.gnome3.gvfs`. XFCE doesn't use a fork of any of these packages do they?

**Comment by worldofpeace** [comment]
> Opening setting dialogs takes visible time

Testing in a VM I also notice this. I think it'll probably be impossible to debug this properly without a module up to date for testing. I think I'll merge this and go finish https://github.com/NixOS/nixpkgs/pull/37594

**Pull request merged**